MazerPlatformer/GameObject.cs

```diff
@@ -162,79 +167,88 @@ public Component AddComponent(Component.ComponentType type, object value, string
+                .Bind(unit => DrawGameObjectBoundingBox(spriteBatch))
+                .Bind(unit=> DrawGameObjectBoundingSphere(spriteBatch));
+
+        // All game objects can ask to draw some text over it if it wants
+        // dependency on Mazer for game font ok.
+        public virtual Either<IFailure, Unit> Draw(SpriteBatch spriteBatch) =>
+            DoIfReturn(!IsNullOrEmpty(InfoText) && Diganostics.DrawObjectInfoText, () => DrawText(spriteBatch))
+                .Bind(unit => DrawObjectDiagnostics(spriteBatch));
+
+        private Either<IFailure, Unit> DrawText(SpriteBatch spriteBatch) => Ensure(() =>
         {
-            DrawCentrePoint(spriteBatch);
-            DrawMaxPoint(spriteBatch);
-            DrawGameObjectBoundingBox(spriteBatch);
-            DrawGameObjectBoundingSphere(spriteBatch);
-        }
-
-        // Specific game objects need to initialize
-        public virtual void Draw(SpriteBatch spriteBatch)
-        {
-            // All game objects can ask to draw some text over it if it wants
-            // dependency on Mazer for game font ok.
-            DoIf(!IsNullOrEmpty(InfoText) && Diganostics.DrawObjectInfoText, () =>
-            {
-                spriteBatch.DrawString(Mazer.GetGameFont(), InfoText, new Vector2(X - 10, Y - 10), Color.White);
-                spriteBatch.DrawString(Mazer.GetGameFont(), SubInfoText ?? string.Empty, new Vector2(X + 10, Y + Height), Color.White);
-            });
-
-            DrawObjectDiagnostics(spriteBatch);
-        }
+            spriteBatch.DrawString(Mazer.GetGameFont(), InfoText, new Vector2(X - 10, Y - 10), Color.White);
+            spriteBatch.DrawString(Mazer.GetGameFont(), SubInfoText ?? string.Empty, new Vector2(X + 10, Y + Height), Color.White);
+        });
 
         // Specific game objects need to initialize themselves
-        public virtual void Initialize()
+        public virtual Either<IFailure, Unit> Initialize() => Ensure(() =>
         {
             // We enter the default state whatever that is
-            foreach(var state in States)
+            foreach (var state in States)
                 StateMachine.AddState(state);
 
-            StateMachine.Initialise(States.Any() 
-                ? States.FirstOrDefault(s=>s.Name == "default")?.Name ?? States.First()?.Name 
+            StateMachine.Initialise(States.Any()
+                ? States.FirstOrDefault(s => s.Name == "default")?.Name ?? States.First()?.Name
                 : null);
-        }
+        });
 
         #endregion
 
         protected virtual void Dispose(bool disposing)
         {
             if (!disposing) return;
 
-            OnDisposing?.Invoke(this);
-
-            // Cleanup objects we know we wont need or that other objects should not need.
-            Components.Clear();
-            States.Clear();
-            StateTransitions.Clear();
+            Ensure(() => OnDisposing?.Invoke(this))
+                .EnsuringMap(unit =>
```

Comment:
Ensuring map will return either a transformation failure or the result of the transformation